### PR TITLE
Address DASC policy review feedback

### PR DESCRIPTION
**Pull request by kaix-nv**
This stacked PR updates #2375 through its protected source branch.

It addresses the automated review findings by:
- making the public state_sparsity API explicit;
- binding exported policy layers, head counts, horizons, masks, and decay identity to the current model;
- keeping checkpoint save/composition non-fatal for stale metadata while preserving strict deployment export;
- allowing recalibration to supersede a stale DASC policy;
- defining the perplexity-retention ratio and accepting valid values above one;
- supporting wrapper unwrapping and exact supported GDN class matching;
- adding actionable generic-mode errors, Pydantic compatibility, documentation, and negative-path tests.

Validation:
- 15 focused DASC tests pass
- 126 combined DASC/sparsity/attention-mode tests pass
- full pre-commit passes
- real Qwen3NextGatedDeltaNet smoke test passes

Signed-off-by: Kai Xu <kaix@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * DASC now supports recognized Gated Delta Network model variants, including wrapped models.
  * Repeated DASC application is supported.
  * Public DASC APIs and policies are more explicitly available.
  * Perplexity retention values above 1 are accepted and documented as valid.

* **Bug Fixes**
  * Improved validation detects unsupported models, malformed measurements, stale policies, and altered model structures.
  * Missing calibration measurements now produce actionable guidance.
  * Stale policies are handled with warnings during refresh and rejected during export.

* **Documentation**
  * Clarified retention metrics, supported adapters, recalibration behavior, and export requirements.

### CODE REVIEW

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 1]
```diff
@@ -58,20 +62,21 @@ def compute_gdn_decay_horizons(
 
 
 def _is_gdn_module(module: nn.Module) -> bool:
-    class_name = "".join(
-        character for character in type(module).__name__.lower() if character.isalnum()
-    )
+    """Return whether a module has one of the explicitly supported GDN implementations."""
     return (
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:66]: Addressed in signed commit cfd7053c3b via #2379. Supported implementations are now recognized through the class MRO, which accepts ModelOpt synthesized `_DynamicGatedDeltaNet` classes because their MRO includes the registered `GatedDeltaNet` base. An unrelated lookalike with `A_log` and `dt_bias` remains rejected. The regression test swaps in a dynamic-style subclass and verifies strict export still succeeds.

File: modelopt/torch/sparsity/state_sparsity/mode.py [hunk 2]
```diff
@@ -48,11 +48,6 @@ def config_class(self) -> type[ModeloptBaseConfig]:
         """Return the validated DASC configuration class."""
         return DASCConfig
 
-    @property
-    def next_prohibited_modes(self) -> set[str]:
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/mode.py:52]: Addressed in signed commit cfd7053c3b via #2379. The repeat-mode prohibition is restored. Public `calibrate()` now detects existing DASC state and replaces that entry in place through `ModeloptStateManager`, updating both config and policy metadata and collapsing any provisional duplicates. The test changes provenance during recalibration and asserts the serialized mode list contains exactly one updated `dasc` entry.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 3]
```diff
@@ -88,19 +96,23 @@ def restore_dasc_model(model: nn.Module, config: DASCConfig, metadata: MetadataD
...
 
 def update_dasc_metadata(model: nn.Module, config: DASCConfig, metadata: MetadataDict) -> None:
-    """Refresh serialized metadata from the immutable attached DASC policy."""
-    try:
-        policy = DASCPolicy(**getattr(model, _DASC_POLICY_ATTRIBUTE))
-    except (AttributeError, TypeError, ValidationError) as error:
-        raise ApplyModeError("Model has no valid attached DASC policy") from error
+    """Refresh metadata without making unrelated ModelOpt save or compose paths unusable."""
+    policy = get_attached_dasc_policy(model)
     validate_dasc_model_structure(model, policy)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py:101]: Addressed in signed commit cfd7053c3b via #2379. Both structure and decay validation now run inside the non-fatal stale-policy handler used by save and composition; `export_policy()` still calls both validators directly and remains strict. A new test changes the GDN head count, verifies `mto.save()` succeeds with a stale-policy warning, and verifies export rejects the structure.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 4]
```diff
@@ -115,11 +122,18 @@ def _model_structure(modules: dict[str, nn.Module]) -> list[dict[str, object]]:
...
-            "dt_bias": module.dt_bias.detach().to(device="cpu", dtype=torch.float64).tolist(),
+            "A_log": module.A_log.detach()
+            .to(device="cpu", dtype=torch.bfloat16)
+            .to(dtype=torch.float32)
+            .tolist(),
+            "dt_bias": module.dt_bias.detach()
+            .to(device="cpu", dtype=torch.bfloat16)
+            .to(dtype=torch.float32)
+            .tolist(),
         }
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:137]: Addressed in signed commit cfd7053c3b via #2379. The exact BF16 digest is now explicitly calibration provenance, not the deployment-validity gate. Export re-derives numerical horizons with cast tolerance and exact-compares the selected retained-head mask, which is the runtime-relevant contract. The dtype regression is parameterized over both BF16 and FP16.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 5]
```diff
@@ -24,11 +24,15 @@
...
 
 from modelopt.torch.opt.conversion import ApplyModeError
+from modelopt.torch.utils import unwrap_model
 
 from .config import DASCCalibrationMeasurement, DASCConfig, DASCLayerPolicy, DASCPolicy
 
 __all__ = ["analyze_gdn_decay", "compute_gdn_decay_horizons"]
 
+_SUPPORTED_GDN_CLASS_NAMES = frozenset({"GatedDeltaNet", "Qwen3NextGatedDeltaNet"})
+_HORIZON_DTYPE_CAST_RTOL = 0.05
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:34]: Addressed by the validation change in signed commit cfd7053c3b via #2379. The exact BF16 digest no longer dominates export validation, so the 5 percent horizon tolerance now genuinely absorbs ordinary FP16/BF16 storage casts. The constant comment states that the exact retained-head mask is the semantic gate, and both dtype paths are covered by tests.

File: modelopt/torch/sparsity/state_sparsity/config.py [hunk 6]
```diff
@@ -67,6 +72,8 @@ def validate_unique_slices(
 class DASCConfig(ModeloptBaseConfig):
     """Configuration for GDN decay-aware state checkpoint sparsity."""
 
+    model_config = ConfigDict(extra="forbid", validate_assignment=True, protected_namespaces=())
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/config.py:75]: Addressed in signed commit cfd7053c3b via #2379. Both subclasses now set only `ConfigDict(protected_namespaces=())`, so `extra` and `validate_assignment` continue to inherit from `ModeloptBaseConfig`.
